### PR TITLE
Fixing errors on develop branch

### LINKED ISSUE #35167: salt v2016.3.2 changes grains['osfinger'] from 'Ubuntu-14.04' to 'Ubuntu-14'
### Description of Issue/Question
Salt versions 2016.3.1 and earlier had Ubuntu 14.04 systems report "Ubuntu-14.04" in the osfinger grain. Salt-minion 2016.3.2 on a similar system will have just "Ubuntu-14" in the same grain.

A system administrator may quite often write jinja templates that check for an exact string in that grain. The change in the grain will require a large number of adjustments in the Jinja templates.

### Setup
No special setup needed. Just some Ubuntu 14.04 minions and a master to control them. My testing was done with the minion and master being the same system.

### Steps to Reproduce Issue
 * Install salt-minion 2016.3.2 on a system running Ubuntu 14.04.
 * Run `salt -G osfinger:Ubuntu-14 grains.item saltversion`

### Versions Report
```
Salt Version:
           Salt: 2016.3.2
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 4.4.0-31-generic
         system: Linux
        version: Ubuntu 14.04 trusty
 ```

### LINKED ISSUE #35219: salt develop broken - state run showing traceback and raising AttributeError
Salt develop branch is broken. When trying to run any state using salt from develop branch, getting the following error:

```
root@nmadhok.local:~/salt/salt# salt-call state.sls sample.state test=true
[WARNING ] Key 'file_ignore_glob' with value None has an invalid type of NoneType, a list or basestring is required for this value
[WARNING ] Key 'file_ignore_regex' with value None has an invalid type of NoneType, a list or basestring is required for this value
[INFO    ] Loading fresh modules for state activity
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
AttributeError: 'bool' object has no attribute 'get'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 352, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 58, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 134, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 197, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 985, in sls
    high_, errors = st_.render_highstate({saltenv: mods})
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 3149, in render_highstate
    sls, saltenv, mods, matches)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2803, in render_state
    state_data = self.client.get_state(sls, saltenv)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 394, in get_state
    dest = self.cache_file(path, saltenv, cachedir=cachedir)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 177, in cache_file
    return self.get_url(path, '', True, saltenv, cachedir=cachedir)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 475, in get_url
    url, dest, makedirs, saltenv, cachedir=cachedir)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 972, in get_file
    hash_server, stat_server = self.hash_and_stat_file(path, saltenv)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 1282, in hash_and_stat_file
    stat_result = fnd.get('stat')
AttributeError: 'bool' object has no attribute 'get'
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 352, in salt_call
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/call.py", line 58, in run
    caller.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 134, in run
    ret = self.call()
  File "/usr/lib/python2.6/site-packages/salt/cli/caller.py", line 197, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.6/site-packages/salt/modules/state.py", line 985, in sls
    high_, errors = st_.render_highstate({saltenv: mods})
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 3149, in render_highstate
    sls, saltenv, mods, matches)
  File "/usr/lib/python2.6/site-packages/salt/state.py", line 2803, in render_state
    state_data = self.client.get_state(sls, saltenv)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 394, in get_state
    dest = self.cache_file(path, saltenv, cachedir=cachedir)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 177, in cache_file
    return self.get_url(path, '', True, saltenv, cachedir=cachedir)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 475, in get_url
    url, dest, makedirs, saltenv, cachedir=cachedir)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 972, in get_file
    hash_server, stat_server = self.hash_and_stat_file(path, saltenv)
  File "/usr/lib/python2.6/site-packages/salt/fileclient.py", line 1282, in hash_and_stat_file
    stat_result = fnd.get('stat')
AttributeError: 'bool' object has no attribute 'get'
```

Looks like the commit https://github.com/saltstack/salt/commit/73a156d28cafd45dcc6c2aac9f1ac3ad905695e3 made by @terminalmage caused it

PR with the fix coming soon!

### LINKED ISSUE #35223: Core grains broken on salt develop branch
Some of the core grains do not populate properly after changes in https://github.com/saltstack/salt/commit/bc73db071a84a61b5711018266ba1d1599096c56 were implemented. 

One of the  broken grain is ``osmajorrelease`` whose type has been changed from ``string`` to ``int`` causing the following state to fail:

State:
```
root@nmadhok.local:/srv/salt# cat sample-state.sls 
{% set osMajorRelease = grains['osmajorrelease'][0] %}

{% if grains['os_family'] == "RedHat" and osMajorRelease == '6'  %}
echo 'I am RHEL6':
  cmd.run
{% endif %}
```

Error:
```
[WARNING ] Key 'file_ignore_glob' with value None has an invalid type of NoneType, a list or basestring is required for this value
[WARNING ] Key 'file_ignore_regex' with value None has an invalid type of NoneType, a list or basestring is required for this value
[INFO    ] Loading fresh modules for state activity
[INFO    ] Fetching file from saltenv 'base', ** done ** 'sample-state.sls'
[ERROR   ] Rendering exception occurred: Jinja variable int object has no element 0
[CRITICAL] Rendering SLS 'base:sample-state' failed: Jinja variable int object has no element 0
[ERROR   ] Data passed to highstate outputter is not a valid highstate return: {'local': ["Rendering SLS 'base:sample-state' failed: Jinja variable int object has no element 0"]}
local:
    - Rendering SLS 'base:sample-state' failed: Jinja variable int object has no element 0
```

Other grains that broke include ``osfinger``: See related issue https://github.com/saltstack/salt/issues/35167

### PR DESCRIPTION
**Pull request by nmadhok**
Fix broken core grains:
Resolves #35223 
Resolves #35167

### CODE REVIEW

File: salt/fileclient.py [hunk 1]
```diff
@@ -1278,8 +1278,13 @@ def hash_and_stat_file(self, path, saltenv='base'):
                 'saltenv': saltenv,
                 'cmd': '_file_find'}
         fnd = self.channel.send(load)
-        stat_result = fnd.get('stat')
-        return hash_result, stat_result
+
+        try:
+            fnd_mode = fnd.get('stat', [])[0]
```
Comment (terminalmage) [inline salt/fileclient.py]: Nope. This won't work. The fileclient expects the list-ified stat_result and checks the first element in the list.

Comment (terminalmage) [inline salt/fileclient.py]: Working on a proper fix.

**Comment by terminalmage** [comment]
Please remove a825b34 from this pull request. While it may have gotten rid of the traceback, it is not the correct fix and will in fact break the new ``- mode: keep`` feature in file.recurse/file.managed states.

#35224 provides the correct fix for #35219, and makes a couple other corrections.

Thanks for the heads-up on this edge case. I would appreciate more information (as requested) in the issue comments for (#35219), I'm curious what would have caused the response to be a boolean rather than just an empty dictionary. I did write a couple of integration tests for this feature but given this edge case I think I ought to write one to cover it.

**Comment by nmadhok** [comment]
@terminalmage Done.

**Comment by isbm** [comment]
@nmadhok I am not a fan of this spaghetti approach. All I see is that Ubuntu is for some really weird reasons is different than Debian in this case and therefore is reporting os_finger differently. I just tested it and just found it true. But I would suggest to always write a unit test for every issue, so to see it is failing. And only _then_ write a code, which would not require this voluminous change.

**Comment by isbm** [comment]
@cachedout :warning:  Please wait with this PR, I am preparing another one with unit tests for Ubuntu.

**Comment by nmadhok** [comment]
@isbm The type of ``osmajorrelease`` was also changed from string to integer. It didn't take me long to put this together since I referenced the code that existed earlier. What surprises me is that changes to core grains were merged in a stable release without any tests and deprecation warnings. It's not just the grain on Ubuntu that broke but also the ``osmajorrelease`` grain on all distributions making the latest stable salt version that we have installed in our environment useless.

**Comment by isbm** [comment]
@nmadhok The integer part was fixed, apparently. The unit test is actually expects a string.

**Comment by nmadhok** [comment]
@isbm https://github.com/saltstack/salt/commit/a6bcbd615fff3f985c4d6b9d9e87aa384142fa3f#diff-6e894b39b67c4bab67796752744ea9d5R1492 No deprecation warnings? Next release this will be an integer? I thought the process for deprecating old code and warnings was atleast after 2 major releases which is when they can be safely removed?

**Comment by isbm** [comment]
@nmadhok fine with me to add a deprecation warning and postpone to the next release. In the meanwhile I confirm the Ubuntu bug and fixing it.

**Comment by nmadhok** [comment]
@isbm Sounds good.

**Comment by isbm** [comment]
@nmadhok Fixed, pretty small change was required. Adding unit test now. PR soon.
